pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,292 @@
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HeadObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// Config Azure storage configuration.
+type Config struct {
+	StorageAccountName string `yaml:"storage-account"`
+	StorageAccountKey  string `yaml:"storage-account-key"`
+	ContainerName      string `yaml:"container"`
+}
+
+// Bucket implements the store.Bucket interface against s3-compatible APIs.
+type Bucket struct {
+	logger       log.Logger
+	containerURL blob.ContainerURL
+	config       *Config
+	opsTotal     *prometheus.CounterVec
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() error {
+	if conf.StorageAccountName == "" ||
+		conf.StorageAccountKey == "" {
+		return errors.New("invalid Azure storage configuration")
+	}
+	return nil
+}
+
+// NewBucket returns a new Bucket using the provided Azure config.
+func NewBucket(logger log.Logger, azureConfig []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
+
+	level.Debug(logger).Log("msg", "Creating new Azure bucket connection", "component", component)
+
+	var conf Config
+	if err := yaml.Unmarshal(azureConfig, &conf); err != nil {
+		return nil, err
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	container, err := createContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+	if err != nil {
+		ret, ok := err.(blob.StorageError)
+		if !ok {
+			return nil, errors.Wrapf(err, "Azure API return unexpected error: %T\n", err)
+		}
+		if ret.ServiceCode() == "ContainerAlreadyExists" {
+			level.Debug(logger).Log("msg", "Getting connection to existing Azure blob container", "container", conf.ContainerName)
+			container, err = getContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot get existing Azure blob container: %s", container)
+			}
+		} else {
+			return nil, errors.Wrapf(err, "error creating Azure blob container: %s", container)
+		}
+	} else {
+		level.Info(logger).Log("msg", "Azure blob container successfully created", "address", container)
+	}
+
+	bkt := &Bucket{
+		logger:       logger,
+		containerURL: container,
+		config:       &conf,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_azure_storage_operations_total",
+			Help:        "Total number of operations that were executed against an Azure storage account.",
+			ConstLabels: prometheus.Labels{"storage_account": conf.StorageAccountName},
+		}, []string{"operation"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
```

Comment:
In future we can move it to more streamed version with go routines. But for now I guess easy version will work..

See this for example: https://github.com/improbable-eng/thanos/pull/467/files#diff-34b0a4c283913d869fe9f2998bafb23aR249 Maybe not the best example as this go routine function is too big and quite complex, but something like this. But only if data will prove that this is problematic.

Reply:
I will investigate how to make it more optimal